core/shared/src/main/scala/spire/math/interval/Overlap.scala

```diff
@@ -0,0 +1,65 @@
+package spire.math.interval
+
+import spire.algebra.{Eq, Order}
+import spire.math.Interval
+import spire.math.interval.Overlap.{Disjoint, Equal, Subset}
+
+/**
+  * An ADT that represents overlapping result for any two intervals.
+  */
+sealed abstract class Overlap[A: Order] extends Serializable {
```

Comment:
Also, super minor, but we should also extend `Product` here (eg `extends Product with Serializable`) as per http://underscore.io/blog/posts/2015/06/04/more-on-sealed.html

Not a big deal - I think we should probably do this throughout Spire, so if it isn't done now, it'll probably happen in a later PR anyways.